ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs

```diff
@@ -311,20 +307,26 @@ txSubmissionInbound _tracer maxUnacked mpReader mpWriter _version =
             -- Check if having received more txs we can now confirm any (in
             -- strict order in the unacknowledgedTxIds sequence).
             (acknowledgedTxIds, unacknowledgedTxIds') =
-              Seq.spanl (`Map.member` bufferedTxs') (unacknowledgedTxIds st)
+              Seq.spanl (`Map.member` btxsReq) (unacknowledgedTxIds st)
 
             -- If so we can submit the acknowledged txs to our local mempool
-            txsReady = foldr (\txid r -> maybe r (:r) (bufferedTxs' Map.! txid))
+            txsReady = foldr (\txid r -> maybe r (:r) (btxsReq Map.! txid))
                              [] acknowledgedTxIds
 
             -- And remove acknowledged txs from our buffer
-            bufferedTxs'' = foldl' (flip Map.delete)
-                                   bufferedTxs' acknowledgedTxIds
+            btxsAcked = foldl' (flip Map.delete)
+                                   btxsReq acknowledgedTxIds
+
+        _writtenTxids <- mempoolAddTxs txsReady
 
-        _ <- mempoolAddTxs txsReady
+        -- If we are acknowleding transactions that are still in unacknowledgedTxIds'
+        -- we need to re-add them so that we also can acknowledge them again later.
+        -- This will happen incase of duplicate txids within the same window.
```

Comment:
> duplicate txids within the same window.

Is there a simple-enough example you could record in this code comment of how this might happen? I don't recall; it's been too long since I touched that repro we had a while ago.

Reply:
This is the simplest example that quickcheck could generate. I don't think it is concise enough to put it in a code comment.
Consider the following list of TXs and a window size greater than or equal to five.
[Tx 0, Tx 1, Tx 2, Tx 3, Tx 2]

 1. The server requests 3 new TxIds.
 2. The client responds with [TxId 0, TxId 1, TxId 2].
 3. The server requests [Tx 0, Tx 1].
 4. The server acks [TxId 0, TxId 1] and request 2 new TxIds.
 5. The client responds with [Tx 0, Tx 1]
 6. The server reqest Tx 2.
 7. The client responds with [TxId 3, TxId 2 (dupe)].
 8. The client responds with Tx 2.
 9. The server request Tx 3
10. The server acks [TxId 2, TxId 3] and request two new TxIds. Note that only the TxId 2 from msg 2 is acked, not the dupe TxId2 from msg 7.
11. The client responds with no more Txids.
12. The server will trigger an assertion failure (line 199) because it has failed to acknowledge the 2 (dupe) TxId.

Reply:
Thanks, that's helpful.

Hmm. Could the comment say "duplicate txids can arise merely as a consequence of pipelining the current BF protocol"? Do you think that is an accurate summary?

Reply:
In the example above pipelining permitted the duplicate txids to trigger an assertion, but pipelining didn't create the duplicate Tx id. The client decided to send a duplicate Tx id within the same window. I think in theory this could happen in case the client sent us a new TxId, rolled back the a fork without that Tx, fetched the Tx again from another peer and then sent us the same TxId again.

It is also possible to trigger this bug if the Txs aren't fetched in order (they are fetched based on `Ord` TxId, not in the order they where presented), in away that doesn't depend on pipelining.